DataViews: add `in the past` and `over` operators for the date field type

chihsuan opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Closes [ARC-218](https://linear.app/a8c/issue/ARC-218/dataviews-date-filter-add-in-the-past-and-overoperators)

## Proposed Changes

Add new `In the past` and `Over` operators for datetime field types in DataViews.

Users can now filter date fields by a relative time period (e.g., "in the past 7 days", "over 1 year ago").

WA Figma design:

<img width="265" alt="Screenshot 2025-06-23 at 17 03 56" src="https://github.com/user-attachments/assets/ae951932-c149-4b1c-8944-7555a8e92fd0" />

hdK9XLiOefugy2bxalKBN1-fi-2806_161863

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

This change introduces more flexible and user-friendly date filtering in DataViews, allowing users to:

- Find items that occurred within a recent period (e.g., "orders in the past 30 days").
- Find items that are older than a certain period (e.g., "Customer joined for over 6 months").

This improves reporting, analytics, and workflow use cases where relative date queries are needed.


## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

- Run `yarn dataviews:storybook:start`
- Go to http://localhost:57863/?path=/story/dataviews-fieldtypes--date-time
- Select the `In the past` or `Over` operator.
- Enter a value and select a unit (days, weeks, months, years).
- Verify that the filtered results match the expected date range.


<img width="292" alt="Screenshot 2025-06-23 at 17 00 53" src="https://github.com/user-attachments/assets/ed5595d2-c192-4a96-8554-51faeda9e018" />
<img width="295" alt="Screenshot 2025-06-23 at 17 00 41" src="https://github.com/user-attachments/assets/f047c8c5-abd5-4702-a1f3-c805c2d99879" />



https://github.com/user-attachments/assets/55f2be33-92d6-4b2f-82fe-45b4d795c56c


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

chihsuan committed:
DataViews: Add new date filtering operators for relative date queries

- Introduced `inThePast` and `over` operators for filtering date fields.
- Updated relevant components and constants to support new operators.
- Enhanced the DateTime control to handle new filtering options.
- Added tests to validate the functionality of the new operators.

chihsuan committed:
Add changelog

chihsuan committed:
Improve readme

chihsuan committed:
Improve variable naming

chihsuan commented on packages/dataviews/src/components/dataviews-filters/filter.tsx:
```diff
@@ -363,16 +389,25 @@ function OperatorSelector( {
+													) ||
+														OPETATORS_SHOULD_RESET_VALUE.includes(
+															operator
+														) );
```
Idea for Future Improvement:

Instead of maintaining a hardcoded list of operators that require value resets when switching, I think we could define a validation function for each filter operator. When the operator changes, we would validate the current value against the new operator’s requirements. If the value is not valid, we reset it to undefined. Something we could consider when we work on [DOTCOM-13677](https://linear.app/a8c/issue/DOTCOM-13677/dataviews-refactor-filtersortandpaginate-function)

oandregal replied:
Resetting could work this way:

- reset if current or new operator is `between`
- reset when current operator is `in-the-past` or `over`, but the new operator is not one of them

This would keep the value when switching from `in-the-past` to `over`.

chihsuan replied:
Yeah, and the value should also be reset when switching from any operator that is not in the past or over to one that is.

This seems overly complex. I'll take a more straightforward approach and enhance it later by adding operator validation logic as mentioned above.

chihsuan committed:
Simplify filter logic

chihsuan committed:
Remove unused useRealTimers

oandregal commented on packages/dataviews/src/components/dataviews-filters/filter.tsx:
```diff
@@ -363,16 +389,25 @@ function OperatorSelector( {
-													operator ===
-														OPERATOR_BETWEEN;
+												// Reset the value only when switching between operators that have different value types.
+												const OPETATORS_SHOULD_RESET_VALUE =
```
typo: `opeRators_should_reset_value`

chihsuan replied:
Thanks, fixed typo in b27a6d5e95bc937e947b6e2026bab73dc080fc27.

oandregal commented on packages/dataviews/src/dataform-controls/datetime.tsx:
```diff
@@ -1,19 +1,110 @@
+	options: { value: string; label: string }[];
+} ) {
+	const { value: relValue = '', unit = options[ 0 ].value } =
+		value && typeof value === 'object' ? value : {};
```
Small thing: this check (returning empty object if value is not an object) could be part of the `RelativeDateControls` instance, like in the `TimePicker`. That way, we can set value's type to `object`.

chihsuan replied:
Good idea. I've updated the code to ensure that the value prop passed to RelativeDateControls is always an object aafd95233ee84ae2603ca2093231772bb5ba3e16

chihsuan committed:
Fix typo

chihsuan committed:
Refactor DateTime component to improve value handling

- Updated the value prop type to be more explicit, allowing for an object with optional properties.
- Simplified value extraction logic in the RelativeDateControls function.
- Ensured consistent handling of value in the DateTime component.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/17573086

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @chihsuan for including a screenshot in the description! This is really helpful for our translators.
